### DISCUSSION TITLE
Writing a qupath-friendly pyramid with Z>1

### DISCUSSION THREAD

**Discussion by marcoaaz** [libvips/libvips#2446] [Help wanted]
Hi, 

I am glad this community is active and I hope I can make a question (more than an enhancement request) around here. I have been recently learning to use pyvips. I would like to know how to write a tiff (or ome.tif) pyramid with various Z or T levels. My data in particular has 7 z-levels (each RGB), which is interpreted in ImageJ as 21 channels (greyscaled). 

My starting point was this [script](https://forum.image.sc/t/writing-qupath-bio-formats-compatible-pyramidal-image-with-libvips/51223/5) from J. Cupitt and I suppose the problem comes down to let QuPath know about the [SUBIFD](https://www.libvips.org/2020/06/18/What's-new-in-8.10.html#pyramids) tag from the metadata. I wrote the respective python [script](https://github.com/marcoaaz/czi-to-readable-ome-tiff-/blob/main/czi_czifile_memory3.py) that I am running from Anaconda prompt producing 21 channels QuPath-friendly pyramids that I do not want. 

For easier understanding, this is a ROI of my [data](https://drive.google.com/file/d/1zrafOa98KiSPKVhRH_bEJH_myGEDAIyW/view?usp=sharing) and I would like to produce a pyramidal image with a shape of CZT = 3x7x1 to be precise. If this is an easy fix or an upcoming feature, please let me know. 

Thank you very much.

**Comment by jcupitt** [comment]
Hello @marcoaaz,

I'm not an expert on OME-TIFF, but I think all you need to do (I think!) is to describe your 21 channel TIFF as 7x3 in the XML metadata.

You need to build the XML in python somehow, then attach it to the image using pyvips as `"image-description"`, then save as ome-tiff.

**Reply by marcoaaz** [reply]
Dear Jhon, I am Marco again. This issue is not resolved for me yet. In the header of the OME-TIFF that you kindly explained [here](https://forum.image.sc/t/writing-qupath-bio-formats-compatible-pyramidal-image-with-libvips/51223/8), the output that I read using Fiji bioformat metadata reader ends up having double fields for each added item (e.g.: SizeX, SizeY, SizeT, and so on..). I read the OME-TIFF documentation and understood how is designed but the python script is probably badly written. Maybe there is some luck and you encountered the same issue before. 

Best wishes,
Marco

**Reply by jcupitt** [reply]
Perhaps you can find a sample OME TIFF with similar layout an adapt their XML for your needs? I think that's what I'd try next.

**Reply by jcupitt** [reply]
I had a quick look at your script and it seems OK, though you could make it a bit simpler and faster.

You're only attaching the minimal XML. You'll need something fancier for 7x3x1, so I don't think this is a libvips issue.

**Reply by marcoaaz** [reply]
Dear Jhon, thanks for the tip. I found the OME-TIFF header from a QuPath export after translating the *.vsi file. The header looks different in the sense that it defines the IFDs individually. In the Python script, I processed *.czi files. In the new script, I will process .vsi but I haven't managed to run Bioformats in Python because the library javabridge "module is not found" after pip installing from Spyder IDE. I am confident that when I figure out why and fix it, writing the header will be fine and QuPath will read the image as I want.

I have a little conceptual question. It seems to me that OME-TIFF has the file as a large matrix (5-d) that is read differently according to the header. If that is the case, I could use bfconvert (BioFormat command line) in a filename sequence and concatenate all *.vsi in an *.ome.tiff (multichannel). Then, overwrite the header to specify the number of Z planes. In your opinion, is this possible?

The problem could also be approached as trying to stack image pyramids in the Z dimension, without ever loading the entire image into the memory. I have found Peter Bankhead's scripts to stack spatially and in channel direction, see [discussion](https://forum.image.sc/t/qupath-multiple-channel-in-separate-files-how-to-merge-them/29455/31).

No need to respond!. I will update, if I find something useful. Have a good day. Cheers,
Marco

**Reply by marcoaaz** [reply]
Dear Jhon, I am updating as I said. The code for producing the pyramids is at this [link](https://github.com/marcoaaz/OlympusVSI-to-OME-TIFF) in my repo. I have documented myself much more and the two options managed by the Image.sc Forum are [here](https://forum.image.sc/t/qupath-bioformats-high-compression-efficiency/58982/3). I hope this is useful to others.

To conclude, I could not go through Libvips in Python because the Bioformat library could not be installed easily, which is my bad. For the time being, this script should give some flexibility to RAM-limited users with an intermediate step where image planes are stored in the HDD (for convenience). 

Best,
Marco.
